### ISSUE TITLE
Secp256k1 failed to parse public key

### ISSUE THREAD

**Issue by deweller** [Bit-Wasp/bitcoin-php#215]
When I try to extract the address from the public key in an input script, Secp256k1 fails to parse it.

``` php
// $transaction is a BitWasp\Bitcoin\Transaction\Transaction
foreach ($transaction->getInputs() as $input) {
    $script = $input->getScript();
    $classifier = new BitWasp\Bitcoin\Script\Classifier\InputClassifier($script);
    if ($classifier->isPayToPublicKeyHash()) {
        $decoded = $script->getScriptParser()->decode();
        $hex_buffer = $decoded[1]->getData();
        $public_key = PublicKeyFactory::fromHex($hex_buffer);
        // Exception thrown: Secp256k1 failed to parse public key
    }
}
```

Is this because Secp256k1 doesn't handle a compressed public key properly?

**Comment by deweller** [comment]
FYI: The EcAdapter works for me with this same bit of code.

**Comment by deweller** [comment]
Actually, on further inspection, it looks like EcAdapter is parsing the publick key incorrectly as well.  It just doesn't throw an error like Secp256k1 does.

**Comment by deweller** [comment]
Here is a simpler example:

```php
$hex_string = '028e92b049f311f02eacc8d3634c40ac9fc9bc2540a2d6d081c7e23d41e6f95358';
$public_key = BitWasp\Bitcoin\Key\PublicKeyFactory::fromHex($hex_string);
$address = $public_key->getAddress()->getHash();
echo "address: $address\n";
```

Actual Result: `0db1f4b2cf314e013be8bf0af8e7e626c744c50b`
Expected: `12FQy5q5Lde1KhWRXz8Hi8HrVjXYNM9njW`

**Comment by afk11** [comment]
Hrm, that is a strange failure! I just tried parsing that key and it was successful. Do the bitcoin-php tests pass? Maybe try the secp256k1-php one's also. 

What PHP version are you using? I'll try a fresh install of libsecp256k1, perhaps they've changed something. 

```
// print_r($public_key)
BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Key\PublicKey Object
(
    [ecAdapter:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Key\PublicKey:private] => BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Adapter\EcAdapter Object
        (
            [math:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Adapter\EcAdapter:private] => BitWasp\Bitcoin\Math\Math Object
                (
                    [compact:BitWasp\Bitcoin\Math\Math:private] => BitWasp\Bitcoin\Math\CompactInteger Object
                        (
                            [math:BitWasp\Bitcoin\Math\CompactInteger:private] => BitWasp\Bitcoin\Math\Math Object
 *RECURSION*
                        )

                )

            [generator:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Adapter\EcAdapter:private] => Mdanter\Ecc\Primitives\GeneratorPoint Object
                (
                    [x] => 55066263022277343669578718895168534326250603453777594175500187360389116729240
                    [y] => 32670510020758816978083085130507043184471273380659243275938904335757337482424
                    [z] => 115792089237316195423570985008687907852837564279074904382605163141518161494337
                    [curve] => Mdanter\Ecc\Curves\NamedCurveFp Object
                        (
                            [a] => 0
                            [b] => 7
                            [prime] => 115792089237316195423570985008687907853269984665640564039457584007908834671663
                        )

                )

            [context:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Adapter\EcAdapter:private] => Resource id #45
        )

    [compressed:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Key\PublicKey:private] => 1
    [pubkey_t:BitWasp\Bitcoin\Crypto\EcAdapter\Impl\Secp256k1\Key\PublicKey:private] => Resource id #50
)
address: 12FQy5q5Lde1KhWRXz8Hi8HrVjXYNM9njW

```

**Comment by deweller** [comment]
The secp256k1-php tests fail for me.  I upgraded to php 5.6.16, but they still fail for me.

```
# php -v
PHP 5.6.16 (cli) (built: Dec  4 2015 08:02:22)
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Techn


PHPUnit 4.8.9 by Sebastian Bergmann and contributors.

.............................................................   61 / 1781 (  3%)
.............................................................  122 / 1781 (  6%)
.............................................................  183 / 1781 ( 10%)
.............................................................  244 / 1781 ( 13%)
.............................................................  305 / 1781 ( 17%)
.............................................................  366 / 1781 ( 20%)
.............................................................  427 / 1781 ( 23%)
.............................................................  488 / 1781 ( 27%)
.............................................................  549 / 1781 ( 30%)
.............................................................  610 / 1781 ( 34%)
.............................................................  671 / 1781 ( 37%)
.............................................................  732 / 1781 ( 41%)
.............................................................  793 / 1781 ( 44%)
.............................................................  854 / 1781 ( 47%)
.............................................................  915 / 1781 ( 51%)
.............................................................  976 / 1781 ( 54%)
............................................................. 1037 / 1781 ( 58%)
............................................................. 1098 / 1781 ( 61%)
............................................................. 1159 / 1781 ( 65%)
............................................................. 1220 / 1781 ( 68%)
..EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE 1281 / 1781 ( 71%)
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE....EEEEEEEEEEEEEEEEE 1342 / 1781 ( 75%)
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE 1403 / 1781 ( 78%)
EEEEEEEEEEEEEEEEEEEEE........................................ 1464 / 1781 ( 82%)
............................................................. 1525 / 1781 ( 85%)
............................................................. 1586 / 1781 ( 89%)
............................................................. 1647 / 1781 ( 92%)
............................................................. 1708 / 1781 ( 95%)
............................................................. 1769 / 1781 ( 99%)
............

Time: 1.22 seconds, Memory: 26.75Mb

There were 198 errors:

1) BitWasp\Secp256k1Tests\Secp256k1PubkeyTweakAddTest::testAddsToPubkey with data set #0 (resource(1056) of type (secp256k1_context), '041a2756dd506e45a1142c7f7f03a...866e79', '01010101010101010101010101010...010101', '04347aa1a74cf333bdc2866c42e47...f5cbbb', 0)
secp256k1_ec_pubkey_tweak_add() expects parameter 2 to be resource, string given

/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakAddTest.php:64
/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakAddTest.php:49

....

99) BitWasp\Secp256k1Tests\Secp256k1PubkeyTweakAddTest::testAddsToPubkey with data set #98 (resource(1056) of type (secp256k1_context), '041a2756dd506e45a1142c7f7f03a...866e79', '63636363636363636363636363636...636363', '04b015e495c5db2270c28a3386d79...0c53f9', 0)
secp256k1_ec_pubkey_tweak_add() expects parameter 2 to be resource, string given

/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakAddTest.php:64
/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakAddTest.php:49

100) BitWasp\Secp256k1Tests\Secp256k1PubkeyTweakMulTest::testMultipliesByPubkey with data set #0 (resource(1056) of type (secp256k1_context), '041a2756dd506e45a1142c7f7f03a...866e79', '01010101010101010101010101010...010101', '04980ec845debde3671926ba83f92...60fd0b', 0)
secp256k1_ec_pubkey_tweak_mul(): supplied argument is not a valid secp256k1_pubkey resource

/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakMulTest.php:64
/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakMulTest.php:49

...

198) BitWasp\Secp256k1Tests\Secp256k1PubkeyTweakMulTest::testMultipliesByPubkey with data set #98 (resource(1056) of type (secp256k1_context), '041a2756dd506e45a1142c7f7f03a...866e79', '63636363636363636363636363636...636363', '0498bbe03889c056b7d707483f55e...11949b', 0)
secp256k1_ec_pubkey_tweak_mul(): supplied argument is not a valid secp256k1_pubkey resource

/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakMulTest.php:64
/work/php-extensions/secp256k1-php/tests/Secp256k1PubkeyTweakMulTest.php:49

FAILURES!
Tests: 1781, Assertions: 4893, Errors: 198.
```

**Comment by afk11** [comment]
Not sure what it could be, but been putting off a refactoring.. I'll look into this tomorrow!

**Comment by afk11** [comment]
Are you running xdebug btw? I'm on 5.6 as well FWIW.

I installed Ubuntu 15 a week ago and am getting errors/segfaults when I run with xdebug. I normally use this stuff over the CLI.. I just tried a script over apache, and got the error from secp256k1 about not parsing the public key. but the tests pass for me over the cli :/

**Comment by deweller** [comment]
I'm using Mac OS X.  The PHP CLI is built using homebrew.

**Comment by afk11** [comment]
I think this might be back working again - would you mind testing the latest version of secp256k1-php? It's mentioned in the linked issue, I couldn't reproduce it without xdebug, but it seems the error goes away by changing the type for pubkeylen in the bindings.

**Comment by afk11** [comment]
Just FYI, the library and extension now run on PHP7. I'm pretty sure the bug leading to this issue was fixed in 5.6, but I'll be testing for similar problems with 7. Let me know if you get to test things!

**Comment by afk11** [comment]
I'm quite sure this error has been addressed, please re-open if you run into it again!
